### PR TITLE
[FLINK-27931][sql-gateway] Introduce the SqlGateway to assemble all components

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Add the main entry for the SqlGateway that assembles all components.*

## Brief change log

  - *Add the SqlGateway to assemble all components.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added the test to verify whether can load the specified endpoint and startup.*
  - *Added the test to verify the print help message*
  - *Added the test to verify to specify duplicate identifiers.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.concurrent.CountDownLatch;
+
+/** Main entry point for the SQL Gateway. */
+public class SqlGateway {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SqlGateway.class);
+
+    private final CountDownLatch latch;
+
+    private SessionManager sessionManager;
```
Comment (lsyldliu) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:50]: Adjust it after `dynamicConfig `?

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        DefaultContext context =
+                DefaultContext.load(ConfigurationUtils.createConfiguration(dynamicConfig));
+        sessionManager = new SessionManager(context);
+
+        sessionManager.start();
+        SqlGatewayService sqlGatewayService = new SqlGatewayServiceImpl(sessionManager);
+
+        Configuration conf = context.getFlinkConfig();
+        endpoints.addAll(
+                SqlGatewayEndpointFactoryUtils.createSqlGatewayEndpoint(sqlGatewayService, conf));
```
Comment (lsyldliu) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:71]:
```suggestion
                SqlGatewayEndpointFactoryUtils.createSqlGatewayEndpoint(sqlGatewayService, context.getFlinkConfig()));
```

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            sessionManager.stop();
+        }
+    }
+
+    public static void main(String[] args) {
+        startSqlGateway(System.out, args);
+    }
+
+    @VisibleForTesting
+    public static void startSqlGateway(PrintStream stream, String[] args) {
```
Comment (lsyldliu) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:99]:
```suggestion
   static void startSqlGateway(PrintStream stream, String[] args) {
```

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    private final Properties dynamicConfig;
+
+    public SqlGateway(Properties dynamicConfig) {
+        endpoints = new ArrayList<>();
+        latch = new CountDownLatch(1);
+
+        this.dynamicConfig = dynamicConfig;
+    }
+
+    public void start() {
```
Comment (lsyldliu) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:61]: Introducing a awaitTerminal api of Endpoint, when anyone of endpoint exit, the gateway also should exit.

Comment (fsk119) [reply flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:56]: Ok. I have already implemented this feature.

File: flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    public void stop() {
+        for (SqlGatewayEndpoint endpoint : endpoints) {
+            stopEndpointSilently(endpoint);
+        }
+        if (sessionManager != null) {
+            sessionManager.stop();
+        }
+    }
+
+    public static void main(String[] args) {
```
Comment (wuchong) [inline flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/SqlGateway.java:88]: I think we can follow `org.apache.flink.runtime.taskexecutor.TaskManagerRunner#main` to log some important environment information.

**Review by wuchong — COMMENTED** [review]
Thanks for the contribution @fsk119 . I left some comments.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request closed without merging**